File: rust/aiconfigurator-core/src/models/deepseek_v4.rs
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! DeepSeek-V4 family op-list builder.
//!
//! Apple-to-apple port of `aiconfigurator.sdk.models.deepseek_v4.DeepSeekV4Model`.
//! Differs from base DeepSeek in three structural ways:
//!
//! - The attention block is the DSv4 compressed-attention module
//!   (`Op::Dsv4Context` / `Op::Dsv4Generation`) instead of MLA.
//!   `compress_ratios` is a per-layer list of {0, 4, 128} where 4 selects
//!   the CSA (compressed-sparse) variant and 128 selects HCA
//!   (hybrid-causal). Pure SWA layers (ratio=0) are approximated using
//!   the HCA tables — matching Python's
//!   `ratio_counts[128] += ratio_counts.pop(0, 0)` mapping. One attention
//!   op per unique ratio with `scale_factor = layer count`.
//! - Each decoder block has a manifold-constrained hyper-connection (mHC)
//!   pair (`mhc_pre` + `mhc_post`) replacing the standard add_norm
//!   elementwise. Currently routed through `Op::Mhc(MhcModuleOp)` which
//!   does not differentiate the `pre`/`post` slice — the Rust MHC table
//!   doesn't carry that key, and no smoke-set perf data exists for DSv4
//!   on b200/h200/h100 systems anyway, so the op-graph mirrors Python's
//!   structure even though numeric values for MHC are approximate when
//!   data does land.
//! - Generation overlaps the routed-MoE chain with the shared expert FFN
//!   (`Op::Overlap`), same idiom as DSv32.
//!
//! `n_shared_experts` is fixed at 1 by Python's default; the shared FFN
//! emits one set of gate_up + act_gate + ffn2 ops.

use std::collections::BTreeMap;

use crate::common::enums::{BackendKind, GemmQuantMode};
use crate::common::error::AicError;
use crate::models::base::{Model, ModelConfig};
use crate::models::config_loader::DeepSeekV4Config;
use crate::operators::{
    DispatchFlavor, Dsv4ModuleOp, ElementwiseOp, EmbeddingOp, GemmOp, MhcModuleOp, MoEDispatchOp,
    MoeOp, Op, OverlapOp, P2POp,
};
use crate::perf_database::dsv4::AttnKind;

fn dispatch_flavor(backend: BackendKind) -> DispatchFlavor {
    match backend {
        BackendKind::Trtllm => DispatchFlavor::TrtllmAlltoall,
        BackendKind::Sglang => DispatchFlavor::CustomAllReduce,
        BackendKind::Vllm => DispatchFlavor::CustomAllReduce,
    }
}

/// Map `compress_ratio` to the `AttnKind` whose perf table to query.
/// Mirrors Python's `ratio_counts[128] += ratio_counts.pop(0, 0)` rule:
/// ratio 0 (pure SWA) is approximated with HCA.
fn attn_kind_for_ratio(ratio: u32) -> AttnKind {
    if ratio == 4 {
        AttnKind::Csa
    } else {
        AttnKind::Hca
    }
}

pub fn build_deepseek_v4_model(config: ModelConfig) -> Result<Model, AicError> {
    let dsv4: DeepSeekV4Config = config
        .spec
        .deepseek_v4
        .as_ref()
        .ok_or_else(|| {
            AicError::UnsupportedModel(
                "DEEPSEEKV4 builder requires deepseek_v4 extras in ModelSpec".to_string(),
            )
        })?
        .clone();
    let mut model = Model::new(config);
    let cfg = &model.config;
    let layers = cfg.spec.num_hidden_layers as f64;
    // Python `_mtp_scale_factor` (models/base.py:105-110); see deepseek.rs
    // for the full rationale. `nextn=0` reduces both to identity.
    let mtp_scale = cfg.mtp_scale_factor();
    let gen_layers = layers * mtp_scale;
    let h = cfg.spec.hidden_size;
    let tp = cfg.parallel.tp_size.max(1);
    let pp = cfg.parallel.pp_size.max(1);
    let vocab_per_tp = cfg.spec.vocab_size / tp;
    let dtypes = cfg.dtypes;
    let num_experts = cfg.spec.num_experts.max(1);
    let topk = cfg.spec.top_k.max(1);
    let moe_inter = cfg.spec.moe_intermediate_size;
    let moe_tp = cfg.parallel.moe_tp_size.max(1);
    let moe_ep = cfg.parallel.moe_ep_size.max(1);
    let attn_dp = cfg.parallel.attention_dp_size.max(1);
    let backend = cfg.backend;
    let flavor = dispatch_flavor(backend);
    let architecture = cfg.spec.architecture.clone();
    let local_heads = cfg.spec.num_attention_heads / tp;
    let moe_inter_per_tp = (moe_inter + tp - 1) / tp;

    // Python: attn_norm / ffn_norm use ElementWise(num_layers, h, h, 0.8) =>
    // (h + h) * 2 bytes/token for bf16.
    let attn_norm_bytes = (h as f64 + h as f64) * 2.0;
    let shared_act_gate_bytes = (2.0 * moe_inter_per_tp as f64 + moe_inter_per_tp as f64) * 2.0;

    // Count compress_ratios per Python's `Counter`. ratio 0 -> 128 (HCA approx).
    let mut ratio_counts: BTreeMap<u32, u32> = BTreeMap::new();
    for &r in &dsv4.compress_ratios {
        *ratio_counts.entry(r).or_insert(0) += 1;
    }
    let swa_count = ratio_counts.remove(&0).unwrap_or(0);
    if swa_count > 0 {
        *ratio_counts.entry(128).or_insert(0) += swa_count;
    }

    let push_attention = |ops_out: &mut Vec<Op>, is_context: bool| {
        let prefix = if is_context { "context" } else { "generation" };
        let name = format!("{prefix}_attention");
        for (&ratio, &count) in &ratio_counts {
            if count == 0 {
                continue;
            }
            let kind = attn_kind_for_ratio(ratio);
            let mut op = Dsv4ModuleOp::new(
                name.clone(),
                kind,
                local_heads,
                dtypes.kv_cache_quant,
                dtypes.fmha_quant,
                dtypes.gemm_quant,
                architecture.clone(),
            );
            op.scale_factor = count as f64;
            ops_out.push(if is_context {
                Op::Dsv4Context(op)
            } else {
                Op::Dsv4Generation(op)
            });
        }
    };

    // Helper to emit one mHC op (pre or post). The Rust MHC perf table
    // doesn't differentiate by phase, so both calls hit the same lookup;
    // Python's phase split surfaces when phase-keyed data is actually
    // present (no smoke-set perf data covers this yet).
    let push_mhc = |ops_out: &mut Vec<Op>, name: &str, scale: f64| {
        let mut op = MhcModuleOp::new(name, dsv4.hc_mult, h, architecture.clone());
        op.scale_factor = scale;
        ops_out.push(Op::Mhc(op));
    };

    // ---- Context phase ----
    let mut ctx = Vec::new();
    ctx.push(Op::Embedding({
        let mut e = EmbeddingOp::new("context_embedding", vocab_per_tp, h, dtypes.gemm_quant);
        e.scale_factor = 0.3;
        e
    }));
    push_mhc(&mut ctx, "context_mhc_pre", layers);
    ctx.push(Op::Elementwise({
        let mut e = ElementwiseOp::new("context_attn_norm", attn_norm_bytes);
        e.scale_factor = layers;
        e
    }));
    push_attention(&mut ctx, true);
    push_mhc(&mut ctx, "context_mhc_post", layers);
    ctx.push(Op::Elementwise({
        let mut e = ElementwiseOp::new("context_ffn_norm", attn_norm_bytes);
        e.scale_factor = layers;
        e
    }));
    // Shared expert (gated FFN).
    ctx.push(Op::Gemm({
        let mut g = GemmOp::new(
            "context_shared_gate_up_gemm",
            2 * moe_inter_per_tp,
            h,
            dtypes.gemm_quant,
        );
        g.scale_factor = layers;
        g
    }));
    ctx.push(Op::Elementwise({
        let mut e = ElementwiseOp::new("context_shared_act_gate", shared_act_gate_bytes);
        e.scale_factor = layers;
        e
    }));
    ctx.push(Op::Gemm({
        let mut g = GemmOp::new(
            "context_shared_ffn2_gemm",
            h,
            moe_inter_per_tp,
            dtypes.gemm_quant,
        );
        g.scale_factor = layers;
        g
    }));
    // Router.
    ctx.push(Op::Gemm({
        let mut g = GemmOp::new(
            "context_router_gemm",
            num_experts,
            h,
            GemmQuantMode::Bfloat16,
        );
        g.scale_factor = layers;
        g
    }));
    // MoE dispatch + compute + dispatch.
    ctx.push(Op::MoeDispatch({
        let mut d = MoEDispatchOp::new(
            "context_moe_pre_dispatch",
            h,
            topk,
            num_experts,
            moe_tp,
            moe_ep,
            attn_dp,
            true,
            backend,
            flavor,
        );
        d.scale_factor = layers;
        d
    }));
    ctx.push(Op::Moe({
        let mut m = MoeOp::new(
            "context_moe",
            h,
            moe_inter,
            topk,
            num_experts,
            moe_tp,
            moe_ep,
            dtypes.moe_quant,
            "power_law_1.01",
        );
        m.scale_factor = layers;
        m
    }));
    ctx.push(Op::MoeDispatch({
        let mut d = MoEDispatchOp::new(
            "context_moe_post_dispatch",
            h,
            topk,
            num_experts,
            moe_tp,
            moe_ep,
            attn_dp,
            false,
            backend,
            flavor,
        );
        d.scale_factor = layers;
        d
    }));
    ctx.push(Op::Gemm(GemmOp::new(
        "context_logits_gemm",
        vocab_per_tp,
        h,
        GemmQuantMode::Bfloat16,
    )));

    // ---- Generation phase ----
    let mut gen = Vec::new();
    gen.push(Op::Embedding({
        let mut e = EmbeddingOp::new("generation_embedding", vocab_per_tp, h, dtypes.gemm_quant);
        e.scale_factor = 0.3 * mtp_scale;
        e
    }));
    push_mhc(&mut gen, "generation_mhc_pre", gen_layers);
    gen.push(Op::Elementwise({
        let mut e = ElementwiseOp::new("generation_attn_norm", attn_norm_bytes);
        e.scale_factor = gen_layers;
        e
    }));
    push_attention(&mut gen, false);
    push_mhc(&mut gen, "generation_mhc_post", gen_layers);
    gen.push(Op::Elementwise({
        let mut e = ElementwiseOp::new("generation_ffn_norm", attn_norm_bytes);
        e.scale_factor = gen_layers;
        e
    }));
    // Overlap: shared FFN ‖ routed MoE.
    let mut gen_shared = Vec::with_capacity(3);
    gen_shared.push(Op::Gemm({
        let mut g = GemmOp::new(
            "generation_shared_gate_up_gemm",
            2 * moe_inter_per_tp,
            h,
            dtypes.gemm_quant,
        );
        g.scale_factor = gen_layers;
        g
    }));
    gen_shared.push(Op::Elementwise({
        let mut e = ElementwiseOp::new("generation_shared_act_gate", shared_act_gate_bytes);
        e.scale_factor = gen_layers;
        e
    }));
    gen_shared.push(Op::Gemm({
        let mut g = GemmOp::new(
            "generation_shared_ffn2_gemm",
            h,
            moe_inter_per_tp,
            dtypes.gemm_quant,
        );
        g.scale_factor = gen_layers;
        g
    }));

    let mut gen_routed = Vec::with_capacity(4);
    gen_routed.push(Op::Gemm({
        let mut g = GemmOp::new(
            "generation_router_gemm",
            num_experts,
            h,
            GemmQuantMode::Bfloat16,
        );
        g.scale_factor = gen_layers;
        g
    }));
    gen_routed.push(Op::MoeDispatch({
        let mut d = MoEDispatchOp::new(
            "generation_moe_pre_dispatch",
            h,
            topk,
            num_experts,
            moe_tp,
            moe_ep,
            attn_dp,
            true,
            backend,
            flavor,
        );
        d.scale_factor = gen_layers;
        d
    }));
    gen_routed.push(Op::Moe({
        let mut m = MoeOp::new(
            "generation_moe",
            h,
            moe_inter,
            topk,
            num_experts,
            moe_tp,
            moe_ep,
            dtypes.moe_quant,
            "power_law_1.01",
        );
        m.scale_factor = gen_layers;
        m
    }));
    gen_routed.push(Op::MoeDispatch({
        let mut d = MoEDispatchOp::new(
            "generation_moe_post_dispatch",
            h,
            topk,
            num_experts,
            moe_tp,
            moe_ep,
            attn_dp,
            false,
            backend,
            flavor,
        );
        d.scale_factor = gen_layers;
        d
    }));
    gen.push(Op::Overlap(OverlapOp::new(
        "generation_moe_overlap",
        gen_routed,
        gen_shared,
    )));
    gen.push(Op::Gemm({
        // Logits GEMM fires once per forward; Python uses `1 * mtp_scale`.
        let mut g = GemmOp::new(
            "generation_logits_gemm",
            vocab_per_tp,
            h,
            GemmQuantMode::Bfloat16,
        );
        g.scale_factor = mtp_scale;
        g
    }));

    // P2P: generation side picks up MTP scale (Python `_mtp_scale_factor`
    // multiplies all generation-phase op scales). `nextn=0` reduces this
    // to the un-scaled path.
    let pp_scale = (pp as f64 - 1.0).max(0.0);
    let mut p2p_ctx = P2POp::new("context_p2p", pp, h);
    p2p_ctx.scale_factor = pp_scale;
    ctx.push(Op::P2P(p2p_ctx));
    let mut p2p_gen = P2POp::new("generation_p2p", pp, h);
    p2p_gen.scale_factor = pp_scale * mtp_scale;
    gen.push(Op::P2P(p2p_gen));

    model.context_ops = ctx;
    model.generation_ops = gen;
    Ok(model)
}

File: src/aiconfigurator/sdk/models/deepseek_v4.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from collections import Counter
from typing import ClassVar

import aiconfigurator.sdk.operations as ops
from aiconfigurator.sdk import common
from aiconfigurator.sdk.models.base import BaseModel, register_model
from aiconfigurator.sdk.models.helpers import calc_expectation


@register_model("DEEPSEEKV4")
class DeepSeekV4Model(BaseModel):
    """DeepSeek-V4 model with mHC plus SWA/CSA/HCA compressed attention."""

    _SUPPORTED_COMPRESS_RATIOS: ClassVar[set[int]] = {0, 4, 128}

    @classmethod
    def supports_cp(cls, backend_name: str) -> bool:
        # DSV4 builds its own per-compress-ratio CP ops in ``__init__``
        # (windowed dense + sparse compressed KV gather, ratio-aware
        # attention scaling). SGLang AllGather only -- see
        # LLAMAModel.supports_cp for the vLLM DCP exclusion rationale.
        return backend_name == "sglang"

    @classmethod
    def create(cls, model_info: dict, model_config, backend_name: str) -> BaseModel:
        return cls(
            model_info["topk"],
            model_info["num_experts"],
            model_info["moe_inter_size"],
            model_info["model_path"],
            model_info["model_family"],
            model_info["architecture"],
            model_info["layers"],
            model_info["n"],
            model_info["n_kv"],
            model_info["d"],
            model_info["hidden_size"],
            model_info["inter_size"],
            model_info["vocab"],
            model_info["context"],
            model_config,
            model_info["extra_params"],
            backend_name=backend_name,
        )

    @property
    def activation_hidden_size(self) -> int:
        # DSv4 attention expands Q/O internals, but resident MoE/residual activations use hidden_size.
        return self._hidden_size

    def __init__(self, topk: int, num_experts: int, moe_inter_size: int, *args, backend_name: str = "") -> None:
        super().__init__(*args)

        if not isinstance(self.extra_params, common.DeepSeekV4Config):
            raise TypeError("DeepSeekV4Model requires DeepSeekV4Config extra_params")
        deepseek_v4_cfg = self.extra_params
        self._compress_ratios = deepseek_v4_cfg.compress_ratios
        unknown_ratios = set(self._compress_ratios) - self._SUPPORTED_COMPRESS_RATIOS
        if unknown_ratios:
            raise ValueError(f"Unsupported DeepSeek-V4 compress_ratios: {sorted(unknown_ratios)}")

        assert self.config.attn_width == self.config.moe_tp_size * self.config.moe_ep_size, (
            f"attn_width tp*cp*dp ({self.config.attn_width}) must equal moe_tp_size * moe_ep_size "
            f"({self.config.moe_tp_size * self.config.moe_ep_size})"
        )
        assert num_experts >= self.config.moe_ep_size, f"ep size cannot be larger than num_experts {num_experts}"

        # The "cp>1 requires tp=1 and dp=1" rule is a SGLang backend constraint
        # (not DSV4-specific) and is enforced in ``get_model`` for any model on
        # SGLang. No DSV4-side guard needed.

        self._topk = topk
        self._num_experts = num_experts
        self._moe_inter_size = moe_inter_size
        self._mtp_scale_factor = (
            1.0
            / (1 + calc_expectation(self._nextn, self._nextn_accept_rates))
            * (self._nextn + self._num_layers)
            / self._num_layers
        )
        self._power_law_alpha = 1.01

        h = self._hidden_size
        tp_size = self.config.tp_size
        moe_tp_size = self.config.moe_tp_size
        moe_ep_size = self.config.moe_ep_size
        attention_dp_size = self.config.attention_dp_size
        pp_size = self.config.pp_size
        moe_backend = self.config.moe_backend
        use_megamoe = moe_backend == "megamoe"
        if use_megamoe:
            if backend_name != common.BackendName.sglang.value:
                raise ValueError("DeepSeek-V4 MegaMoE modeling is only supported with the SGLang backend.")
            if moe_tp_size != 1:
                raise ValueError(f"DeepSeek-V4 MegaMoE requires moe_tp_size=1, got {moe_tp_size}.")
            if moe_ep_size <= 1:
                raise ValueError(f"DeepSeek-V4 MegaMoE requires moe_ep_size > 1, got {moe_ep_size}.")
        # CP is gated by ``supports_cp`` in ``get_model`` -- by the time we reach
        # __init__, cp_size > 1 implies a backend that has CP perf data.
        context_cp_size = self.config.cp_size

        gemm_quant_mode = self.config.gemm_quant_mode
        moe_quant_mode = self.config.moe_quant_mode
        kvcache_quant_mode = self.config.kvcache_quant_mode
        fmha_quant_mode = self.config.fmha_quant_mode
        workload_distribution = (
            self.config.workload_distribution + f"_{self._power_law_alpha}"
            if self.config.workload_distribution == "power_law"
            else self.config.workload_distribution
        )
        local_heads = self._num_heads // tp_size
        local_o_groups = max(1, deepseek_v4_cfg.o_groups // tp_size)
        local_moe_inter_size = self._moe_inter_size // tp_size

        # CP (sequence parallelism). DSV4 sglang AllGather only. The
        # ``_attention_ops`` helper below multiplies its returned count by
        # ``scale_factor / attn_count_div`` so the SWA/CSA/HCA modules get
        # ``count /= cp`` directly at construction (mirrors the
        # ``self._num_layers / attn_count_div`` pattern used by other models).
        # Token-major ops get ``seq_split=cp`` inline at construction; the
        # per-ratio CP NCCL gather ops produced by ``_context_cp_all_gather_ops``
        # carry no ``seq_split`` because their cross-rank fan-out is already
        # encoded in ``num_gpus=cp_size``.
        cp_style = self.config.cp_style
        attn_count_div = context_cp_size if cp_style in ("allgather", "ring") else 1
        cp = context_cp_size  # short alias for ``seq_split=cp`` below

        def _attention_ops(is_context: bool, scale_factor: float):
            ratio_counts = Counter(self._compress_ratios)
            # Some DeepSeek-V4 configs include pure SWA layers (compress_ratio=0).
            # Approximate their module latency with HCA (compress_ratio=128) so
            # the model reuses DeepSeek-V4 HCA perf data instead of requiring a
            # dedicated SWA collector. KV cache capacity below still uses the
            # real per-layer ratios.
            ratio_counts[128] += ratio_counts.pop(0, 0)
            op_cls = ops.ContextDeepSeekV4AttentionModule if is_context else ops.GenerationDeepSeekV4AttentionModule
            name = "context_attention" if is_context else "generation_attention"
            return [
                op_cls(
                    name,
                    count * scale_factor,
                    local_heads,
                    self._num_heads,
                    tp_size,
                    h,
                    deepseek_v4_cfg.q_lora_rank,
                    deepseek_v4_cfg.o_lora_rank,
                    deepseek_v4_cfg.head_dim,
                    deepseek_v4_cfg.qk_rope_head_dim,
                    deepseek_v4_cfg.index_n_heads,
                    deepseek_v4_cfg.index_head_dim,
                    deepseek_v4_cfg.index_topk,
                    deepseek_v4_cfg.sliding_window,
                    ratio,
                    local_o_groups,
                    kvcache_quant_mode,
                    fmha_quant_mode,
                    gemm_quant_mode,
                )
                for ratio, count in ratio_counts.items()
                if count > 0
            ]

        def _moe_ops(phase: str, num_layers: float, is_context: bool, seq_split: int = 1):
            if use_megamoe:
                return [
                    ops.DeepSeekV4MegaMoEModule(
                        f"{phase}_megamoe",
                        num_layers,
                        h,
                        self._moe_inter_size,
                        self._topk,
                        self._num_experts,
                        moe_tp_size,
                        moe_ep_size,
                        moe_quant_mode,
                        workload_distribution,
                        is_context=is_context,
                        seq_split=seq_split,
                    )
                ]
            return [
                ops.MoEDispatch(
                    f"{phase}_moe_pre_dispatch",
                    num_layers,
                    h,
                    self._topk,
                    self._num_experts,
                    moe_tp_size,
                    moe_ep_size,
                    attention_dp_size,
                    True,
                    quant_mode=moe_quant_mode,
                    seq_split=seq_split,
                ),
                ops.MoE(
                    f"{phase}_moe",
                    num_layers,
                    h,
                    self._moe_inter_size,
                    self._topk,
                    self._num_experts,
                    moe_tp_size,
                    moe_ep_size,
                    moe_quant_mode,
                    workload_distribution,
                    attention_dp_size,
                ),
                ops.MoEDispatch(
                    f"{phase}_moe_post_dispatch",
                    num_layers,
                    h,
                    self._topk,
                    self._num_experts,
                    moe_tp_size,
                    moe_ep_size,
                    attention_dp_size,
                    False,
                    quant_mode=moe_quant_mode,
                    seq_split=seq_split,
                ),
            ]

        def _context_cp_all_gather_ops(scale_factor: float):
            """NCCL all-gather ops per distinct DSV4 compress ratio.

            DSV4 KV layout per layer is "windowed dense recent + sparse old":
              bytes_per_seq = window * cache_entry_bytes           (constant)
                            + (seq_len / ratio) * cache_entry_bytes (linear)
                            + (seq_len / ratio) * indexer_bytes     (linear, ratio=4 only)

            Per ratio:
              - r=0   (SWA, dense windowed): 1 op, per-token=0 + window_constant per-seq.
              - r=4   (DSA, NSA-style): TWO ops -- the SGLang CP path issues
                separate AllGathers for the indexer key and the latent cache
                (mirrors ``DeepSeekV32Model._cp_attn_comm_ops``). The
                window_constant per-seq term lives on the cache op.
              - r=128 (CSA, compressed): 1 op, cache-only per-token + window_constant per-seq.

            TODO(cp-window-cap): the windowed-dense term is treated as the full
            ``sliding_window * cache_entry_bytes`` constant per seq, which over-
            states traffic when ``seq_len < sliding_window`` (rare under CP, which
            is typically only enabled for long-context prompts). Tightening this
            requires extending the NCCL op to expose a per-seq cap; deferred.
            """
            if context_cp_size <= 1:
                return []
            cache_entry_bytes = deepseek_v4_cfg.head_dim * kvcache_quant_mode.value.memory
            indexer_bytes = common.deepseek_v4_indexer_cache_entry_bytes(deepseek_v4_cfg.index_head_dim)
            comm_bytes = self.config.comm_quant_mode.value.memory
            window_constant_elements = deepseek_v4_cfg.sliding_window * cache_entry_bytes / comm_bytes

            ratio_counts = Counter(self._compress_ratios)
            cp_ops = []
            for ratio, count in ratio_counts.items():
                if count <= 0:
                    continue
                if ratio == 0:
                    # SWA: only the window_constant per-seq term.
                    cp_ops.append(
                        ops.NCCL(
                            f"context_cp_all_gather_r{ratio}",
                            count * scale_factor,
                            "all_gather",
                            num_elements_per_token=0.0,
                            num_gpus=context_cp_size,
                            comm_quant_mode=self.config.comm_quant_mode,
                            num_elements_per_seq=window_constant_elements,
                        )
                    )
                    continue
                cache_per_token_elements = (cache_entry_bytes / ratio) / comm_bytes
                cp_ops.append(
                    ops.NCCL(
                        f"context_cp_all_gather_r{ratio}_cache",
                        count * scale_factor,
                        "all_gather",
                        num_elements_per_token=cache_per_token_elements,
                        num_gpus=context_cp_size,
                        comm_quant_mode=self.config.comm_quant_mode,
                        num_elements_per_seq=window_constant_elements,
                    )
                )
                if ratio == 4:
                    # DSA (NSA): separate indexer-key gather, mirrors DSV3.2.
                    indexer_per_token_elements = (indexer_bytes / ratio) / comm_bytes
                    cp_ops.append(
                        ops.NCCL(
                            f"context_cp_all_gather_r{ratio}_indexer",
                            count * scale_factor,
                            "all_gather",
                            num_elements_per_token=indexer_per_token_elements,
                            num_gpus=context_cp_size,
                            comm_quant_mode=self.config.comm_quant_mode,
                        )
                    )
            return cp_ops

        context_moe_ops = _moe_ops("context", self._num_layers, is_context=True, seq_split=cp)

        self.context_ops.extend(
            [
                ops.Embedding("context_embedding", 1, self._vocab_size, h, 0.3, seq_split=cp),
                ops.DeepSeekV4MHCModule(
                    "context_mhc_pre",
                    self._num_layers,
                    "pre",
                    h,
                    deepseek_v4_cfg.hc_mult,
                    deepseek_v4_cfg.hc_sinkhorn_iters,
                    common.GEMMQuantMode.bfloat16,
                    seq_split=cp,
                ),
                ops.ElementWise("context_attn_norm", self._num_layers, h, h, 0.8, seq_split=cp),
                *_attention_ops(is_context=True, scale_factor=1.0 / attn_count_div),
                *_context_cp_all_gather_ops(scale_factor=1.0),
                ops.DeepSeekV4MHCModule(
                    "context_mhc_post",
                    self._num_layers,
                    "post",
                    h,
                    deepseek_v4_cfg.hc_mult,
                    deepseek_v4_cfg.hc_sinkhorn_iters,
                    common.GEMMQuantMode.bfloat16,
                    seq_split=cp,
                ),
                ops.ElementWise("context_ffn_norm", self._num_layers, h, h, 0.8, seq_split=cp),
                ops.GEMM(
                    "context_shared_gate_up_gemm",
                    self._num_layers,
                    2 * local_moe_inter_size,
                    h,
                    gemm_quant_mode,
                    seq_split=cp,
                ),
                ops.ElementWise(
                    "context_shared_act_gate",
                    self._num_layers,
                    2 * local_moe_inter_size,
                    local_moe_inter_size,
                    0.8,
                    seq_split=cp,
                ),
                ops.GEMM(
                    "context_shared_ffn2_gemm",
                    self._num_layers,
                    h,
                    local_moe_inter_size,
                    gemm_quant_mode,
                    seq_split=cp,
                ),
                ops.GEMM(
                    "context_router_gemm",
                    self._num_layers,
                    self._num_experts,
                    h,
                    common.GEMMQuantMode.bfloat16,
                    seq_split=cp,
                ),
                *context_moe_ops,
                ops.GEMM(
                    "context_logits_gemm",
                    1,
                    self._vocab_size // tp_size,
                    h,
                    common.GEMMQuantMode.bfloat16,
                    seq_split=cp,
                ),
            ]
        )

        self.generation_ops.extend(
            [
                ops.Embedding("generation_embedding", 1 * self._mtp_scale_factor, self._vocab_size, h, 0.3),
                ops.DeepSeekV4MHCModule(
                    "generation_mhc_pre",
                    self._num_layers * self._mtp_scale_factor,
                    "pre",
                    h,
                    deepseek_v4_cfg.hc_mult,
                    deepseek_v4_cfg.hc_sinkhorn_iters,
                    common.GEMMQuantMode.bfloat16,
                ),
                ops.ElementWise("generation_attn_norm", self._num_layers * self._mtp_scale_factor, h, h, 0.8),
                *_attention_ops(is_context=False, scale_factor=self._mtp_scale_factor),
                ops.DeepSeekV4MHCModule(
                    "generation_mhc_post",
                    self._num_layers * self._mtp_scale_factor,
                    "post",
                    h,
                    deepseek_v4_cfg.hc_mult,
                    deepseek_v4_cfg.hc_sinkhorn_iters,
                    common.GEMMQuantMode.bfloat16,
                ),
                ops.ElementWise("generation_ffn_norm", self._num_layers * self._mtp_scale_factor, h, h, 0.8),
            ]
        )

        gen_shared_ops = [
            ops.GEMM(
                "generation_shared_gate_up_gemm",
                self._num_layers * self._mtp_scale_factor,
                2 * local_moe_inter_size,
                h,
                gemm_quant_mode,
            ),
            ops.ElementWise(
                "generation_shared_act_gate",
                self._num_layers * self._mtp_scale_factor,
                2 * local_moe_inter_size,
                local_moe_inter_size,
                0.8,
            ),
            ops.GEMM(
                "generation_shared_ffn2_gemm",
                self._num_layers * self._mtp_scale_factor,
                h,
                local_moe_inter_size,
                gemm_quant_mode,
            ),
        ]
        generation_moe_ops = _moe_ops(
            "generation",
            self._num_layers * self._mtp_scale_factor,
            is_context=False,
        )
        gen_routed_ops = [
            ops.GEMM(
                "generation_router_gemm",
                self._num_layers * self._mtp_scale_factor,
                self._num_experts,
                h,
                common.GEMMQuantMode.bfloat16,
            ),
            *generation_moe_ops,
        ]
        self.generation_ops.append(
            ops.OverlapOp("generation_moe_overlap", group_a=gen_routed_ops, group_b=gen_shared_ops)
        )
        self.generation_ops.append(
            ops.GEMM(
                "generation_logits_gemm",
                1 * self._mtp_scale_factor,
                self._vocab_size // tp_size,
                h,
                common.GEMMQuantMode.bfloat16,
            )
        )

        pp_scale_factor = pp_size - 1
        self.context_ops.append(ops.P2P("context_p2p", pp_scale_factor, h, pp_size, seq_split=cp))
        self.generation_ops.append(ops.P2P("generation_p2p", pp_scale_factor * self._mtp_scale_factor, h, pp_size))

    # ----- KV bytes API: per-ratio sparse layout (windowed + compressed) -----

    def get_kvcache_bytes_per_sequence(self, seq_len: int) -> float:
        deepseek_v4_cfg = self.extra_params
        seq_len = max(0, seq_len)
        total = 0.0
        cache_entry_bytes = deepseek_v4_cfg.head_dim * self.config.kvcache_quant_mode.value.memory
        for ratio in self._compress_ratios:
            total += min(seq_len, deepseek_v4_cfg.sliding_window) * cache_entry_bytes
            if ratio:
                compressed_entries = seq_len // ratio
                total += compressed_entries * cache_entry_bytes
                coff = 2 if ratio == 4 else 1
                # Compressor decode state keeps FP32 kv_state and score_state buffers.
                total += 2 * ratio * coff * deepseek_v4_cfg.head_dim * 4
                if ratio == 4:
                    total += compressed_entries * common.deepseek_v4_indexer_cache_entry_bytes(
                        deepseek_v4_cfg.index_head_dim
                    )
                    # CSA has a second FP4 indexer compressor with its own decode state.
                    total += 2 * ratio * 2 * deepseek_v4_cfg.index_head_dim * 4
        return total
